File: src/main/java/io/github/legacy_fabric_community/serialization/registry/IdentifiableRegistry.java
package io.github.legacy_fabric_community.serialization.registry;

import java.util.Random;
import java.util.Set;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import javax.annotation.Nullable;

import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

public interface IdentifiableRegistry<T> extends Registry<Identifier, T> {
    T getRandom(Random random);

    Set<Identifier> getIds();

    int getRawId(@Nullable T entry);

    @Nullable
    Identifier getId(T entry);

    default Stream<T> stream() {
        return StreamSupport.stream(this.spliterator(), false);
    }
}

File: shadowed-config-registries/src/main/scala/io/github/legacy_fabric_community/shadowedconfig/registries/IdentifiableRegistry.scala
package io.github.legacy_fabric_community.shadowedconfig.registries

import java.util
import java.util.stream.{Stream, StreamSupport}
import java.util.{Optional, Random}
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import org.jetbrains.annotations.Nullable

trait IdentifiableRegistry[T] extends Registry[Identifier, T] {
	def getRandom(random: Random): Optional[T]

	def getIds: util.Set[Identifier]

	def getRawId(@Nullable entry: T): Int

	@Nullable def getId(entry: T): Identifier

	def stream: Stream[T] = StreamSupport.stream(this.spliterator, false)
}
